File: 98-validate-binary-search-tree/ken/98.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """

        def traverse(root):
            left_BST = True
            left_min = root.val
            left_max = root.val
            right_BST = True
            right_min = root.val
            right_max = root.val

            if root.left != None:
                left_BST, left_min, left_max = traverse(root.left)
                if left_BST == False or left_max >= root.val:
                    return [False, left_min, left_max]

            if root.right != None:
                right_BST, right_min, right_max = traverse(root.right)
                if right_BST == False or right_min <= root.val:
                    return [False, right_min, right_max]

            return [True, left_min, right_max]

        if root == None:
            return True

        bst, min, max = traverse(root)

        return bst

File: 98-validate-binary-search-tree/leochang/98.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	bool isValidBST(TreeNode* root) {
		if (root == NULL) return true;
		int minVal, maxVal;
		return isValidBSTIn(root, minVal, maxVal);
	}
private:
	bool isValidBSTIn(TreeNode* root, int &min_in, int &max_in) {
		int leftminVal = root->val, leftmaxVal = root->val;
		int rightminVal = root->val, rightmaxVal = root->val;
		bool leftValid = true, rightValid = true;

		if (root->left != NULL) {
			leftValid = isValidBSTIn(root->left, leftminVal, leftmaxVal);
			if (leftmaxVal >= root->val) return false;
			leftminVal = min(root->val, leftminVal);
			leftmaxVal = max(root->val, leftmaxVal);
		}

		if (root->right != NULL) {
			rightValid = isValidBSTIn(root->right, rightminVal, rightmaxVal);
			if (rightminVal <= root->val) return false;
			rightminVal = min(root->val, rightminVal);
			rightmaxVal = max(root->val, rightmaxVal);
		}

		min_in = min(leftminVal, rightminVal);
		max_in = max(leftmaxVal, rightmaxVal);

		return leftValid && rightValid;
	}
};
